[docs] Add CI/CD tutorial (#44866)
expo/expo: docs/pages/tutorial/cicd/e2e-tests.mdx

<Step label="1">

### Add the e2e-test build profile

Add an `e2e-test` profile to **eas.json**. This profile creates an unsigned APK for Android and a simulator build for iOS:

```json eas.json
{
  /* @hide ... */ /* @end */
  "build": {
    "e2e-test": {
      "withoutCredentials": true,
      "android": {
        "buildType": "apk",
        "image": "latest"
      },
      "ios": {
        "simulator": true,
        "image": "latest"
      }
    }
  }
}
```

</Step>

<Step label="2">

### Add Maestro flow files

Create a **.maestro/** directory at the root of the project and then add the following test files.

For example, the following code snippet adds a basic home screen test that launches the app and checks that the "Welcome" text is visible:

```yaml .maestro/home.yml
appId: com.yourname.yourapp # Replace with your app's package name/bundle identifier from app config file
---
- launchApp
- assertVisible: 'Welcome'
```

Then, we can add another test for the settings screen:

```yaml .maestro/navigate.yml

> appId: com.yourname.yourapp # Replace with your app's package name/bundle identifier from app config file

Reviewer: Can we link to the reference for which field in app config is the app ID?

---
title: 'Run E2E tests with Maestro on EAS Workflows'
sidebar_title: E2E tests
description: Learn how to automate end-to-end (E2E) tests with Maestro using Android and iOS development builds on EAS Workflows.
---

import { ContentSpotlight } from '~/ui/components/ContentSpotlight';
import { ProgressTracker } from '~/ui/components/ProgressTracker';
import { Terminal } from '~/ui/components/Snippet';
import { Step } from '~/ui/components/Step';

> Broken navigation flows and crashing screens often slip past code review because reviewers cannot run the app on every pull request.

Reviewer 2: The opener states the problem and stops just before the bullets. A resolution that mirrors it gives readers the payoff up front:

```suggestion
Broken navigation flows and crashing screens often slip past code review because reviewers cannot run the app on every pull request. By the end of this chapter, E2E tests run the app automatically on every pull request and catch those bugs before they merge.
```

[Maestro](https://maestro.mobile.dev/) runs automated flows against a built app on an Android Emulator or an iOS Simulator. It can tap buttons, type text, assert screen content, and navigate between screens. EAS Workflows runs these flows in the cloud after building an APK for Android or an iOS Simulator build.

## Set up E2E tests

For Android and iOS, we need to add a special [build profile](/build/eas-json/) for E2E tests. This profile skips credential setup for Android and allows running our iOS build inside a simulator.

<Step label="1">

### Add the e2e-test build profile

Add an `e2e-test` profile to **eas.json**. This profile creates an unsigned APK for Android and a simulator build for iOS:

```json eas.json
{
  /* @hide ... */ /* @end */
  "build": {
    "e2e-test": {
      "withoutCredentials": true,
      "android": {
        "buildType": "apk",
        "image": "latest"
      },
      "ios": {
        "simulator": true,
        "image": "latest"
      }
    }
  }
}
```

</Step>

<Step label="2">

### Add Maestro flow files

> Create a **.maestro/** directory at the root of the project and then add the following test files.

Reviewer 2: Same monorepo-clarity tweak as in earlier chapters:

```suggestion
Create a **.maestro/** directory at the root of the Expo project and then add the following test files.
```

> For example, the following code snippet adds a basic home screen test that launches the app and checks that the "Welcome" text is visible. Replace `com.yourname.yourapp` in `appId` with your app's [`android.package`](/versions/latest/config/app/#package) on Android or [`ios.bundleIdentifier`](/versions/latest/config/app/#bundleidentifier) on iOS, both defined in the **app.json**:

Author: **[suggestion]** Voice break. The rest of this tutorial set consistently uses `we`/`our`, but this prose sentence switches to `your app's` (`expo-docs-style-guide.md#voice-and-tone`, tutorial exception). The same `your app's` repeats inside the YAML comment on lines 68 and 77, but comments inside code blocks more naturally use 2nd person and are usually fine.

```suggestion
For example, the following code snippet adds a basic home screen test that launches the app and checks that the "Welcome" text is visible. Replace `com.yourname.yourapp` in `appId` with our app's [`android.package`](/versions/latest/config/app/#package) on Android or [`ios.bundleIdentifier`](/versions/latest/config/app/#bundleidentifier) on iOS, both defined in the **app.json**:
```

<Terminal cmd={['$ eas workflow:run .eas/workflows/e2e-tests.yml']} />

On the EAS dashboard, notice that the Android and iOS builds start in parallel. Once each build completes, the corresponding Maestro test job runs the flow files against it.

<ContentSpotlight
  alt="EAS Workflows dashboard showing a Maestro E2E test run with parallel Android and iOS test jobs on development builds."
  src="/static/images/tutorial/cicd/eas-workflows-maestro-e2e-run.png"
  className="max-w-160"
/>

<ContentSpotlight
  alt="EAS Workflows dashboard showing successful Maestro E2E test jobs completed for Android and iOS development builds."
  src="/static/images/tutorial/cicd/eas-workflows-maestro-e2e-success.png"
  className="max-w-160"
/>

</Step>

## Trigger E2E tests automatically

So far we've run the workflow manually. We can trigger it from GitHub events so it runs on every pull request or on demand with a pull request label.

### Run on every pull request

To run E2E tests on every pull request, add an `on.pull_request` trigger to the workflow:

```yaml .eas/workflows/e2e-tests.yml
on:
  pull_request:
    branches: ['*']
```

### Run on demand with a label

> To run E2E tests only when needed, use the [`on.pull_request_labeled`](/eas/workflows/syntax/#onpull_request_labeled) trigger. It accepts a `labels` parameter, and the workflow only runs when one of those labels is added to a pull request. For example, we can use the `test` label to trigger our E2E tests workflow:

Author: **[suggestion]** Passive voice (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
To run E2E tests only when needed, use the [`on.pull_request_labeled`](/eas/workflows/syntax/#onpull_request_labeled) trigger. It accepts a `labels` parameter, and the workflow only runs when someone adds one of those labels to a pull request. For example, we can use the `test` label to trigger our E2E tests workflow:
```

> To run E2E tests only when needed, use the [`on.pull_request_labeled`](/eas/workflows/syntax/#onpull_request_labeled) trigger. It accepts a `labels` parameter, and the workflow will only run when one of those labels is added to a pull request. For example, we can use the `test` label to trigger our E2E tests workflow:

Author: **[nit]** Present tense (`expo-docs-style-guide.md#voice-and-tone`).

```suggestion
To run E2E tests only when needed, use the [`on.pull_request_labeled`](/eas/workflows/syntax/#onpull_request_labeled) trigger. It accepts a `labels` parameter, and the workflow only runs when one of those labels is added to a pull request. For example, we can use the `test` label to trigger our E2E tests workflow:
```
